pkg/certificates/certificates.go

```diff
@@ -0,0 +1,286 @@
+type CertificateSigner struct {
+	Log *zap.Logger
+}
+
+// AuthorizationDB stores authorizations which may be claimed in exchange for a
+// certificate signature.
+type AuthorizationDB struct {
+	DB storage.KeyValueStore
+}
+
+// Authorizations is a slice of authorizations for convenient de/serialization
+// and grouping.
+type Authorizations []*Authorization
+
+// Authorization represents a single-use authorization token and it's status
+type Authorization struct {
+	Token Token
+	Claim *Claim
+}
+
+// Token is a random byte array to be used like a pre-shared key for claiming
+// certificate signatures.
+type Token [tokenLength]byte
+
+// Claim holds information about the circumstances under which an authorization
+// token was claimed.
+type Claim struct {
+	IP         string
+	Timestamp  int64
+	Identity   *provider.PeerIdentity
+	SignedCert *x509.Certificate
+}
+
+// NewServer creates a new certificate signing grpc server
+func NewServer(log *zap.Logger) pb.CertificatesServer {
+	srv := CertificateSigner{
+		Log: log,
+	}
+
+	return &srv
+}
+
+// NewAuthorization creates a new, unclaimed authorization with a random token value
+func NewAuthorization() (*Authorization, error) {
+	var token Token
+	i, err := rand.Read(token[:])
+	if err != nil {
+		return nil, ErrAuthorization.Wrap(err)
+	}
+	if i != tokenLength {
+		return nil, ErrNotEnoughRandom
+	}
+
+	return &Authorization{
+		Token: token,
+	}, nil
+}
+
+// NewAuthDB creates or opens the authorization database specified by the config
+func (c CertSignerConfig) NewAuthDB() (*AuthorizationDB, error) {
+	// TODO: refactor db selection logic?
+	driver, source, err := utils.SplitDBURL(c.AuthorizationDBURL)
+	if err != nil {
+		return nil, peertls.ErrRevocationDB.Wrap(err)
+	}
+
+	authDB := new(AuthorizationDB)
+	switch driver {
+	case "bolt":
+		if c.Overwrite {
+			if err := os.Remove(source); err != nil {
+				return nil, err
+			}
+		}
+
+		authDB.DB, err = boltdb.New(source, AuthorizationsBucket)
+		if err != nil {
+			return nil, ErrAuthorizationDB.Wrap(err)
+		}
+	case "redis":
+		redisClient, err := redis.NewClientFrom(c.AuthorizationDBURL)
+		if err != nil {
+			return nil, ErrAuthorizationDB.Wrap(err)
+		}
+
+		if c.Overwrite {
+			if err := redisClient.FlushDB(); err != nil {
+				return nil, err
+			}
+		}
+
+		authDB.DB = redisClient
+	default:
+		return nil, ErrAuthorizationDB.New("database scheme not supported: %s", driver)
+	}
+
+	return authDB, nil
+}
+
+// Run implements the responsibility interface, starting a certificate signing server.
+func (c CertSignerConfig) Run(ctx context.Context, server *provider.Provider) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	srv := NewServer(zap.L())
+	pb.RegisterCertificatesServer(server.GRPC(), srv)
+
+	return server.Run(ctx)
+}
+
+// Sign signs a valid certificate signing request's cert.
+func (c CertificateSigner) Sign(ctx context.Context, req *pb.SigningRequest) (*pb.SigningResponse, error) {
```

Comment:
Should I assume `Sign()` will be updated in another PR?

Reply:
Correct